Components/LearningCore/Source/gov/sandia/cognition/learning/algorithm/clustering/OptimizedAffinityPropagation.java

```diff
@@ -0,0 +1,318 @@
+     *         divergence between two examples.
+     * @param  selfDivergence The maxValue for self-divergence to use, which
+     *         controls the number of clusters created.
+     * @param  dampingFactor The damping factor (lambda). Must be between 0.0
+     *         and 1.0.
+     * @param  maxIterations The maximum number of iterations.
+     */
+    OptimizedAffinityPropagation(
+        DivergenceFunction<? super T, ? super T> divergence,
+        double selfDivergence,
+        double dampingFactor,
+        int maxIterations)
+    {
+      super(divergence, selfDivergence, dampingFactor, maxIterations);
+    }
+
+    @Override
+    protected boolean initializeAlgorithm()
+    {
+      if (this.getData() == null || this.getData().isEmpty())
+      {
+        // Make sure that the data is valid.
+        return false;
+      }
+
+      // Initialize the main data for the algorithm.
+      setExamples(new ArrayList<>(getData()));
+      setSimilarities(new double[exampleCount][exampleCount]);
+      setResponsibilities(new double[exampleCount][exampleCount]);
+      setAvailabilities(new double[exampleCount][exampleCount]);
+
+      double[] divergenciesArray = new double[exampleCount * (exampleCount+1)/2];
+      boolean computeSelfDivergence = Double.isNaN(getSelfDivergence());
+
+      // Compute the similarity matrix.
+      fillSimilarityStructures(divergenciesArray);
+
+      if(computeSelfDivergence) {
+        double median = computeMedian(divergenciesArray);
+        setSelfDivergence(median);
+      }
+
+      fillSimilarityDiagonal();
+
+      // Initialize the assignments to -1, the changed count, and the
+      // clusters.
+      this.setAssignments(new int[exampleCount]);
+      this.setChangedCount(exampleCount);
+      this.setClusters(new HashMap<>());
+      for (int i = 0; i < exampleCount; i++)
+      {
+        this.assignments[i] = -1;
+      }
+
+      // Ready to learn.
+      return true;
+    }
+
+    /**
+     * Set the self similarity based on the self divergence.
+     */
+    private void fillSimilarityDiagonal()
+    {
+      for (int i = 0; i < exampleCount; i++)
+      {
+        similarities[i][i] = -getSelfDivergence();
+      }
+    }
+
+    private void fillSimilarityStructures(double[] divergenciesArray)
+    {
+      for (int i = 0; i < exampleCount; i++)
+      {
+        final T exampleI = examples.get(i);
+
+        for (int j = 0; j <= i; j++)
+        {
+          // We compute similarity, which is the negative of divergence,
+          // since a lower divergence means a higher similarity.
+          final T exampleJ = examples.get(j);
+          final double similarity = -divergence.evaluate(
+              exampleI, exampleJ);
+          similarities[i][j] = similarity;
+          similarities[j][i] = similarity;
+          divergenciesArray[i * (i+1) / 2 + j] = -similarity;
+        }
+      }
+    }
+
+    private double computeMedian(double[] divergenciesArray)

```

Comment:
This seems like it could be cleaned up a bit and added to UnivariateStatisticsUtil. (At least add a TODO about it : )

Reply:
I added TODO, sorry